python/ray/ml/train/integrations/huggingface/huggingface_trainer.py

```diff
@@ -146,52 +174,391 @@ def trainer_init_per_worker(train_dataset, eval_dataset, **config):
+                        return model
+                    kwargs = {}
+                    # same logic as in transformers.Trainer
+                    if self.args.ddp_find_unused_parameters is not None:
+                        kwargs[
+                            "find_unused_parameters"
+                        ] = self.args.ddp_find_unused_parameters
+                    elif isinstance(model, transformers.trainer.PreTrainedModel):
+                        # find_unused_parameters breaks checkpointing as per
+                        # https://github.com/huggingface/transformers/pull/4659#issuecomment-643356021
+                        kwargs[
+                            "find_unused_parameters"
+                        ] = not model.is_gradient_checkpointing
+                    else:
+                        kwargs["find_unused_parameters"] = True
+
+                    if self.args.ddp_bucket_cap_mb is not None:
+                        kwargs["bucket_cap_mb"] = self.args.ddp_bucket_cap_mb
+                    return train.torch.prepare_model(model, ddp_kwargs=kwargs)
+
+                def _save(self, *args, **kwargs):
+                    # Workaround for RayTrainingArguments not being
+                    # pickleable due to it being defined in a local
+                    # scope
+                    self.args.__class__ = base_training_arguments_class
+                    ret = super()._save(*args, **kwargs)
+                    self.args.__class__ = RayTrainingArguments
+                    return ret
+
+            trainer.__class__ = RayTrainer
+            trainer.args.__class__ = RayTrainingArguments
+            trainer.args.no_cuda = not torch.cuda.is_available()
+            trainer.args.save_on_each_node = True
+            trainer.add_callback(_TrainReportCallback)
+            if trainer.args.device.type == "cuda":
+                torch.cuda.set_device(trainer.args.device)
+
+            checkpoint = train.load_checkpoint()
+            checkpoint_path = None
+            if checkpoint:
+                source_ip = checkpoint[NODE_IP_KEY]
+                source_path = checkpoint[CHECKPOINT_PATH_ON_NODE_KEY]
+                target_ip = get_node_ip_address()
+                if source_ip == target_ip:
+                    checkpoint_path = source_path
+                else:
+                    # TODO(yard1): Confirm if tempdir is the right approach here.
+                    checkpoint_path = tempfile.mkdtemp()
+                    sync_dir_between_nodes(
+                        source_ip=source_ip,
+                        source_path=source_path,
+                        target_ip=target_ip,
+                        target_path=checkpoint_path,
+                        return_futures=False,
+                        max_size_bytes=None,
+                    )
+            trainer.train(resume_from_checkpoint=checkpoint_path)
 
         return train_loop_per_worker
+
+
+class _HFIterableDatasetWithLen(IterableDataset):
```

Comment:
Some of the private classes/functions should be moved to a separate file.

Reply:
Yes, I'd vote to move most if not all of them and only keep the high level trainer in this file